### PR TITLE
New GPIO input doc

### LINKED ISSUE #31651: New IoT GPIO input doc
This is a gap that needs coverage. Shows usage of `GpioController` class to read input from GPIO pins.

### PR DESCRIPTION
**Pull request by CamSoper**
Fixes #31651 

[Staging link](https://review.learn.microsoft.com/en-us/dotnet/iot/tutorials/gpio-input?branch=pr-en-us-31599)

### CODE REVIEW

File: docs/iot/tutorials/gpio-input.md [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+## Prepare the hardware
+
+Use the hardware components to build the circuit as depicted in the following diagram:
+
+:::image type="content" source="https://via.placeholder.com/400x200.png" alt-text="A Fritzing diagram showing a circuit that connects a ground pin to pin 21." lightbox="https://via.placeholder.com/800x400.png":::
+
+The image above depicts a direct connection between a ground pin and pin 21.
+
+> [!TIP]
+> The diagram uses a breadboard and GPIO breakout for illustrative purposes, but feel free to just jumper a ground pin and pin 21 on the Raspberry Pi.
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:37]: `"but feel free to just jumper a ground pin"`

I know what this is trying to communicate, but I'm not sure this phrasing is accurate. It's kind of hard to read, but I'm not sure how I'd rephrase it.

File: docs/iot/tutorials/gpio-input.md [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+Complete the following steps in your preferred development environment:
+
+1. Create a new .NET Console App using either the [.NET CLI](../../core/tools/dotnet-new.md) or [Visual Studio](../../core/tutorials/with-visual-studio.md). Name it *InputTutorial*.
+
+    ```dotnetcli
+    dotnet new console -o InputTutorial
+    ```
+
+1. [!INCLUDE [tutorial-add-packages](../includes/tutorial-add-packages.md)]
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:51]: I'm pretty sure we're missing a step here to `cd` into the newly created _InputTutorial_ project directory. Consider adding this missing step.

File: docs/iot/tutorials/gpio-input.md [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    - GPIO pin 21 is opened with  `PinMode.InputPullUp`.
+        - This opens the pin with a *PullUp* resistor engaged. In this mode, when the pin is connected to ground, it will return `PinValue.Low`. When the pin is disconnected from ground and the circuit is open, the pin returns `PinValue.High`.
+    - The initial status is written to a console using a ternary expression. The pin's current state is read with `Read()`. If it's `PinValue.High`, it writes the `alert` string to the console. Otherwise, it writes the `ready` string.
+    - `RegisterCallbackForPinValueChangedEvent()` registers an anonymous callback function for both the `PinEventTypes.Rising` and `PinEventTypes.Falling` events on the pin. These events correspond to pin states of `PinValue.High` and `PinValue.Low`, respectively.
+    - The callback function delegates to a method called `onPinEvent`. `onPinEvent` uses another ternary expression that also writes the corresponding `alert` or `ready` strings.
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:63]:
```suggestion
    - The callback function delegates to a method called `OnPinEvent`. `OnPinEvent` uses another ternary expression that also writes the corresponding `alert` or `ready` strings.
```

Consider renaming this as all other C# methods.

File: docs/iot/tutorials/gpio-input.md [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+1. Run the app on the Raspberry Pi by switching to the deployment directory and running the executable.
+
+    ```bash
+    ./InputTutorial
+    ```
+
+    The app displays text similar to the following:
+
+    ```console
+    Initial status (05/10/2022 15:59:25): Ready ✅
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:77]:
```suggestion
    Initial status (05/10/2022 15:59:25): READY ✅
```

File: docs/iot/tutorials/gpio-input.md [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    ```
+
+1. Disconnect pin 21 from ground. The console displays text similar to the following:
+
+    ```console
+    (05/10/2022 15:59:59) ALERT! 🚨
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:83]:
```suggestion
    (05/10/2022 15:59:59) ALERT 🚨
```

File: docs/iot/tutorials/gpio-input.md [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    ```
+
+1. Reconnect pin 21 and ground. The console displays text similar to the following:
+
+    ```console
+    (05/10/2022 16:00:25) Ready ✅
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:89]:
```suggestion
    (05/10/2022 16:00:25) READY ✅
```

File: docs/iot/tutorials/gpio-input.md [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    ```
+
+1. Terminate the program by pressing <kbd>Ctrl</kbd>+<kbd>C</kbd>.
+
+Congratulations! You've used GPIO to detect input! There are many uses for this type of input. This example can be used with any scenario where a switch connects or breaks a circuit. Here's an example using it with a magnetic reed switch, which is often used to detect open doors or windows.
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:94]:
```suggestion
Congratulations, you've used GPIO to detect input using the `Iot.Device.Bindings` NuGet package! There are many uses for this type of input. This example can be used with any scenario where a switch connects or breaks a circuit. Here's an example using it with a magnetic reed switch, which is often used to detect open doors or windows.
```

File: docs/iot/tutorials/gpio-input.md [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+:::image type="content" source="https://via.placeholder.com/600x400.png" alt-text="Animated GIF showing magnetic reed switch":::
+
+## Laser tripwire
+
+A similar application involves laser tripwires. Building a laser tripwire requires the following additional components:
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:100]: Let's transition a bit from the previous example.

```suggestion
Extending the previous example concept a bit further, let's take a look at how this could be applied to creating a laser tripwire. Building a laser tripwire requires the following additional components:
```

File: docs/iot/tutorials/gpio-input.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+* Laser receiver sensor module *(see note below)*
+* 10K Ω resistor
+* 20K Ω resistor
+
+> [!NOTE]
+> *Laser receiver sensor module* is the generic name applied to a common module found at many internet retailers. The device may vary in name or manufacturer, but should resemble this image.
+>
+> ![Image of a Laser Receiver Sensor Module](https://via.placeholder.com/200x200.png)
+
+### Hardware
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:112]: This heading is a bit generic, how about this instead?
```suggestion
### Connect laser tripwire hardware
```

File: docs/iot/tutorials/gpio-input.md [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+
+Connect the components as detailed in following diagram.
+
+:::image type="content" source="https://via.placeholder.com/500x300.png" alt-text="Fritzing diagram":::
+
+Pay special attention to the 10K Ω and 20K Ω resistors. These implement a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider). This is because the laser receiver module outputs 5V to indicate the beam is broken. Raspberry Pi only supports up to 3.3V for GPIO input. Since sending the full 5V to the pin could damage the Raspberry Pi, the current from the receiver module is passed through a voltage divider to halve the voltage to 2.5V.
+
+### Software changes
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:120]:
```suggestion
### Apply source code updates
```

File: docs/iot/tutorials/gpio-input.md [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+>
+> ![Image of a Laser Receiver Sensor Module](https://via.placeholder.com/200x200.png)
+
+### Hardware
+
+Connect the components as detailed in following diagram.
+
+:::image type="content" source="https://via.placeholder.com/500x300.png" alt-text="Fritzing diagram":::
+
+Pay special attention to the 10K Ω and 20K Ω resistors. These implement a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider). This is because the laser receiver module outputs 5V to indicate the beam is broken. Raspberry Pi only supports up to 3.3V for GPIO input. Since sending the full 5V to the pin could damage the Raspberry Pi, the current from the receiver module is passed through a voltage divider to halve the voltage to 2.5V.
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:118]:
```suggestion
Pay close attention to the 10K Ω and 20K Ω resistors. These implement a [voltage divider](https://en.wikipedia.org/wiki/Voltage_divider). This is because the laser receiver module outputs 5V to indicate the beam is broken. Raspberry Pi only supports up to 3.3V for GPIO input. Since sending the full 5V to the pin could damage the Raspberry Pi, the current from the receiver module is passed through a voltage divider to halve the voltage to 2.5V.
```

Also, I heard a few peers mention in the past that we should try to avoid Wikipedia links as the validity is unverified. Instead, consider creating a .NET IoT glossary article, and writing your own interpretation of these sorts of things.

**Review by IEvangelist — APPROVED** [review]
I know that you were specifically asking for thoughts on the flow of this tutorial, I really love it! 💜

It's straightforward and objective-centric, which lends itself well to enabling our .NET IoT customers to quickly get started using the `Iot.Device.Bindings` NuGet 📦 and get something working. Kudos! I do have a few nits, and suggestions, most of which attribute to the flow.

The third section, titled **Prepare the hardware**, is a bit jarring. Perhaps an introductory paragraph would help? Something that lets the reader know that they're going to be connecting circuitry. Are there any general safety concerns about the connectivity, or any special tools recommended for the job, or is it just plug and play? Set the expectations here, and consider a few simple steps that the reader should follow. Does that make sense?

Otherwise, I only have a few other suggestions which are all inlined. 🙌🏽

File: docs/iot/tutorials/gpio-input.md [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+
+### Software changes
+
+You can *almost* use the same code as earlier, with one exception. In the other examples, we used `PinMode.InputPullUp` so that when the pin is disconnected from ground and the circuit is open, the pin returns `PinValue.High`.
+
+However, in the case of the laser receiver module, we're not detecting an open circuit. Instead, we want the pin to act as a sink for current coming from the laser receiver module. In this case, we'll open the pin with `PinMode.InputPullDown`. This way, the pin returns `PinValue.Low` when it's getting no current, and `PinValue.High` when it receives current from the laser receiver module.
```
Comment (IEvangelist) [inline docs/iot/tutorials/gpio-input.md:124]: Consider showing the final version of the code for the laser tripwire.

File: docs/iot/tutorials/gpio-input.md (hunk 3 above)
Comment (CamSoper) [reply docs/iot/tutorials/gpio-input.md:63]: Wait... C# method names should be Pascal Case? 

Why have I been using Camel Case? My entire life is a lie.

Comment (CamSoper) [reply docs/iot/tutorials/gpio-input.md:63]: No, it turns out, I knew that. Then why does it suddenly seem wrong?

I'm scared. Hold me.

File: docs/iot/tutorials/gpio-input.md [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+```
+
+> [!IMPORTANT]
+> Make sure the code deployed on your Raspberry Pi includes this change before testing a laser tripwire. The program *does* work without it, but using the wrong input mode risks damage to your Raspberry Pi!
+
+:::image type="content" source="https://via.placeholder.com/600x400.png" alt-text="Animated GIF showing the laser tripwire":::
+
+## Get the source code
+
+The source for this tutorial is [available on GitHub](https://github.com/MicrosoftDocs/dotnet-iot-assets/tree/main/tutorials/InputTutorial).
```
Comment (BillWagner) [inline docs/iot/tutorials/gpio-input.md:137]: This area is where I'd give readers some ideas for how to apply it. Detecting a switch can do a lot of different things. Just some brainstorming:

- turning on indicators / displays.
- outside lighting / auto dusk-dawn turn on / turn off
- do not disturb
- bring coffee...

The laser is one of those ideas.

**Review by BillWagner — APPROVED** [review]
This looks great. I had a couple thoughts and ideas.

File: docs/iot/tutorials/gpio-input.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+[!INCLUDE [ensure-ssh](../includes/ensure-ssh.md)]
+
+## Prepare the hardware
+
+Use the hardware components to build the circuit as depicted in the following diagram:
+
+:::image type="content" source="../media/rpi-jumper-21-gnd-thumb.png" alt-text="A diagram showing a circuit that connects a ground pin to pin 21." lightbox="../media/rpi-jumper-21-gnd-thumb.png":::
+
+The image above depicts a direct connection between a ground pin and pin 21.
```
Comment (joperezr) [inline docs/iot/tutorials/gpio-input.md:34]: Is it ok to assume people would know which pin is pin 21? I'm not sure how many people are used to the Gpio numbering, there are also different numbering schemes to be aware of, like the board pin numbering or the logical GPIO one.
